packages/cli/src/executions/execution-recovery.service.ts

```diff
@@ -0,0 +1,189 @@
+import Container, { Service } from 'typedi';
+import { Push } from '@/push';
+import { sleep } from 'n8n-workflow';
+import { ExecutionRepository } from '@db/repositories/execution.repository';
+import { getWorkflowHooksMain } from '@/WorkflowExecuteAdditionalData'; // @TODO: Dependency cycle
+import { InternalHooks } from '@/InternalHooks'; // @TODO: Dependency cycle if injected
+import type { DateTime } from 'luxon';
+import type { IRun, ITaskData } from 'n8n-workflow';
+import type { EventMessageTypes } from '../eventbus/EventMessageClasses';
+import type { IExecutionResponse } from '@/Interfaces';
+import { NodeCrashedError } from '@/errors/node-crashed.error';
+import { WorkflowCrashedError } from '@/errors/workflow-crashed.error';
+import { ARTIFICIAL_TASK_DATA } from '@/constants';
+
+/**
+ * Service for recovering executions truncated by an instance crash.
+ */
+@Service()
+export class ExecutionRecoveryService {
+	constructor(
+		private readonly push: Push,
+		private readonly executionRepository: ExecutionRepository,
+	) {}
+
+	/**
+	 * "Recovery" means (1) amending key properties of a truncated execution,
+	 * (2) running post-execution hooks, and (3) returning the amended execution
+	 * so the UI can reflect the error. "Recovery" does **not** mean injecting
+	 * execution data from the logs (they hold none), or resuming the execution
+	 * from the point of truncation, or re-running the whole execution.
+	 *
+	 * Recovery is only possible if event logs are available in the container.
+	 * In regular mode, logs should but might not be available, e.g. due to container
+	 * being recycled, max log size causing rotation, etc. In queue mode, as workers
+	 * log to their own filesystems, only manual exections can be recovered.
+	 */
+	async recover(executionId: string, messages: EventMessageTypes[]) {
+		if (messages.length === 0) return null;
+
+		const amendedExecution = await this.amend(executionId, messages);
+
+		if (!amendedExecution) return null;
+
+		await this.executionRepository.updateExistingExecution(executionId, amendedExecution);
+
+		await this.runHooks(amendedExecution);
+
+		this.push.once('editorUiConnected', async () => {
```

Comment:
Sadly we don't support something like `oncePerClient` yet. I could see how to implement it but likely out of scope for this PR? I Imagine something roughly like this:

```ts
// abstract.push.ts

oncePerClient(eventName: string, pushRef: string, action: () => Promise<void>) {
  const eventId = `${eventName}:${pushRef}`;

  const prevAction = this.onceListeners[eventId];

  if (prevAction) this.removeListener(eventId, prevAction);

  const onceAction = async () => {
    await action();
    delete this.onceListeners[eventId];
  };

  this.onceListeners[eventId] = onceAction;

  this.once(eventId, onceAction);
}
```

Reply:
Hum i'm afraid this could break stuff more than fix wdyt?

If we have multiple tabs or multiple users, the odds of getting it to fire to the correct UI would be greatly reduced.

Maybe we should add this handler and remove it after like 5 seconds.

Reply:
Created a follow-up story: https://linear.app/n8n/issue/PAY-1665/ensure-crash-recovery-pushes-to-all-connected-clients-once